enhancements/automated-resource-request-scaling.md

```diff
@@ -0,0 +1,517 @@
+
+Due to the resource-intensive workload, the overhead for platform
+components such as the OpenShift platform and add-on operators is
+severely restricted when compared to other uses. For example, one
+customer has allocated 4 hyperthreads for all components other than
+their own workloads. OpenShift currently requires 7.
+
+### Goals
+
+* This enhancement describes an approach for configuring OpenShift
+  clusters to run with lower CPU resource requests than its default
+  configuration.
+* Clusters built in this way should pass the same Kubernetes and
+  OpenShift conformance and functional end-to-end tests as single-node
+  deployments that have not been scaled down.
+* We have a goal of 4 hyperthreads today, but we do not know what
+  constraints we might be given in the future so we need a solution
+  that is not tied to the current known limit.
+
+### Non-Goals
+
+* This enhancement does not address memory resource requests.
+* This enhancement does not address ways to disable operators or
+  operands entirely.
+* Although the results of this enhancement may be useful for
+  single-node developer deployments, multi-node production
+  deployments, and Internet-of-things devices, those use cases are not
+  addressed directly in this document.
+
+## Proposal
+
+There are two aspects to manage scaling the control plane resource
+requests. Most of the cluster operators are deployed by the
+cluster-version-operator using static manifests that include the
+resource requests. Those operators then deploy other controllers or
+workloads dynamically. We need to account for both sets of components.
+
+
+
+This is where we get down to the nitty gritty of what the proposal actually is.
+
+### User Stories
+
+Detail the things that people will be able to do if this is implemented.
+Include as much detail as possible so that people can understand the "how" of
+the system. The goal here is to make this feel real for users without getting
+bogged down.
+
+Include a story on how this proposal will be operationalized:  lifecycled, monitored and remediated at scale.
+
+#### Story 1
+
+#### Story 2
+
+### Implementation Details/Notes/Constraints [optional]
+
+What are the caveats to the implementation? What are some important details that
+didn't come across above. Go in to as much detail as necessary here. This might
+be a good place to talk about core concepts and how they relate.
+
+### Risks and Mitigations
+
+What are the risks of this proposal and how do we mitigate. Think broadly. For
+example, consider both security and how this will impact the larger OKD
+ecosystem.
+
+How will security be reviewed and by whom? How will UX be reviewed and by whom?
+
+Consider including folks that also work outside your immediate sub-project.
+
+## Design Details
+
+### Open Questions
+
+1. The cluster-version-operator manages cluster operators using static
+   manifests. How would we make changes to the settings of Deployments
+   defined in those manifests?
```

Comment:
I'm still not convinced that we actually need to tune requests for CVO-managed operators, because operator consumption seems like it would be fairly independent of cluster workload (although I expect operands like etcd member and API-server consumption to scale with cluster workload).  Do we have an example CVO-managed deployment where we have one number we like for the general case and a different number we like for the single-node case?

Reply:
@browsell does have some examples of operators that request more resources than they need in the *RAN* use case. I don't know if changing those requests for all other use cases makes sense.

The console operator requests 50m and uses 5m, when the console itself is disabled.

The insights operator requests 10m and uses 1m, in a disconnected environment.

The full list is a bit long to paste here, but there are several others and all together they add up to requests of 210m where the average use is only a total of 22m.

Reply:
I'm curious about the difference.  I guess there are some API requests they're saving when they only have to monitor their config to notice they're off, and don't have to manage any operands.  Things like Insights I expect to be extremely bursty when they're doing real work, and then sleeping in between, and that's hard to represent with `requests` in a sane way, because you can fit a lot of bursty things in the same space if the bursts don't overlap, but it gets tight if the bursts do overlap.  Maybe it would help with scheduling if the Insights operator launched higher-request jobs when it was configured to run, and then the operator itself could lower its request settings?

Reply:
The _requests_ really do matter, too. If we have to fit into N CPUs and we're requesting N + 0.5 CPUs, trimming the requests down can help squeeze into that space. Then, as you point out, the bursty nature of actual use may cause problems, too, and so we may end up having to solve those on a case-by-case basis. Hence the large matrix of options to consider at the end of the doc.

Reply:
I am not saying requests don't matter.  This enhancement is about finding ways to request OCP components to trim down even more beyond what they'd otherwise want to consume for performance reasons.  But the console/insights examples you gave are just plain over-requesting.  You're not asking the console operator to do less than it wants to do; you're just pushing back on it requesting 50m when in this case it only needs 5m to do everything it wants to do.  So the more situations like that we can find, and make adjustments to get requests in line with actual consumption, the better job the scheduler can do finding folks space, and the fewer places we'll need to tell operators to tone down what they're actually doing.  A console operator managing an active console operand may actually need 50m in most cases; I dunno.  But I'm pretty sure that the bulk of the time, the Insights operator is just sleeping between gathers.  If we can restructure the Insights operator into something with a constant, low CPU consumption, it can lower its requests on all deployments, not just single-node deployments.  And it can use jobs for the periodic, bursty gathers that set higher requests.  And those jobs will never fire on single-node deployments, so those folks don't have to budget CPU for them.  And the jobs will fire for other folks, but maybe we set them low priority, so they only get scheduled into gaps created by other bursty OCP-core workload.  And once we've tuned down that sort of over-request issue, we can come back and get a better feel for how and where we want to ask operators to reduce beyond their current single-node usage.

Reply:
One thought is that the appropriate cpu request value for an operator on an edge deployed single-node cluster may not be the same as for a larger multi-node cluster. The advantage of the global options approach below is that (when limited to infrastructure/system-critical pods) it can be used to help bring the infrastructure requests in line with actual usage without having to update each operator. This would limit the need for modifications to operator behavior to those cases where actual CPU load exceeds the available capacity budgeted for overhead.

Reply:
@wking I think your suggestion is at least partly covered by "Lower the resource requests for cluster operators statically in the default configuration" on line 397. That section does not cover having those operators change the way they manage tasks that require CPU bursts, and what you describe here makes me think a one-size-fits-all approach isn't likely to be as effective, long-term, as more thoughtful design changes. Do you think it's sufficient to extend that section to say that we may need to rewrite some of the bursty operators, or should that be a separate alternative entirely?

@imiller0 Would your idea be covered by either "Use a cluster profile to change the resource requests for control plane operators" or one of the alternatives for scaling dynamically?

Reply:
I would echo the same concern with cluster profiles that you called out below (potential for a lot of customer specific profiles). Dynamic scaling is more along the lines of what I was thinking, with the question being what mechanism/where to do the scaling. Using kubelet to do the scaling or increasing allocatable capacity (as described under the global options) when the pod is scheduled seems to avoid some of the issues with modifying the pod's request and then dealing with reconciliation.

Reply:
@wking & @imiller0, I think I have captured the details from this thread in the latest draft. Please let me know if I've missed anything.

Reply:
Bursty comment you have in the current draft works for me.  I'd still like to see a breakdown of:

1. Current requests.
2. Current consumption under Standard Cluster workloads (whatever we feel like a standard cluster is 🤷).
3. Current consumption under single-node workloads.
4. Ideal consumption under single-node workloads (e.g. with slower polling, or whatever [you expect containers to do][1] to further reduce their consumption below 3).

Before and after this sort of consumption stabilization effort.  Because my naive gut still feels like 2 and 3 are going to end up pretty close together for most containers, in which case once we align 1, we don't need a mechanism for tuning it per-cluster.  But I don't think we need to hold this thread open for that; we can pick it up in [this new thread][2] ;)

[1]: https://github.com/openshift/enhancements/pull/628#discussion_r570578301
[2]: https://github.com/openshift/enhancements/pull/628#discussion_r579631867

Reply:
@browsell does your team have the data @wking is asking for? If you can share it with me, I can update the enhancement.